src/System.Text.JsonLab/System/Text/Json/JsonWriter.cs

```diff
@@ -235,21 +820,230 @@ public void WriteValue(string value)
+
+            _bufferWriter.Advance(bytesNeeded);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int CountDigits(ulong value)
+        {
+            int digits = 1;
+            uint part;
+            if (value >= 10000000)
+            {
+                if (value >= 100000000000000)
+                {
+                    part = (uint)(value / 100000000000000);
+                    digits += 14;
+                }
+                else
+                {
+                    part = (uint)(value / 10000000);
+                    digits += 7;
+                }
+            }
+            else
+            {
+                part = (uint)value;
+            }
+
+            if (part < 10)
+            {
+                // no-op
+            }
+            else if (part < 100)
+            {
+                digits += 1;
+            }
+            else if (part < 1000)
+            {
+                digits += 2;
+            }
+            else if (part < 10000)
+            {
+                digits += 3;
+            }
+            else if (part < 100000)
+            {
+                digits += 4;
+            }
+            else if (part < 1000000)
+            {
+                digits += 5;
+            }
+            else
+            {
+                Debug.Assert(part < 10000000);
+                digits += 6;
+            }
+
+            return digits;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void WriteDigitsUInt64D(ulong value, Span<char> buffer)
+        {
+            // We can mutate the 'value' parameter since it's a copy-by-value local.
+            // It'll be used to represent the value left over after each division by 10.
+
+            Debug.Assert(CountDigits(value) == buffer.Length);
+
+            for (int i = buffer.Length - 1; i >= 1; i--)
+            {
+                ulong temp = '0' + value;
+                value /= 10;
+                buffer[i] = (char)(temp - (value * 10));
+            }
+
+            Debug.Assert(value < 10);
+            buffer[0] = (char)('0' + value);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void WriteDigitsUInt64D(ulong value, Span<byte> buffer)
+        {
+            // We can mutate the 'value' parameter since it's a copy-by-value local.
+            // It'll be used to represent the value left over after each division by 10.
+
+            Debug.Assert(CountDigits(value) == buffer.Length);
+
+            for (int i = buffer.Length - 1; i >= 1; i--)
+            {
+                ulong temp = '0' + value;
+                value /= 10;
+                buffer[i] = (byte)(temp - (value * 10));
+            }
+
+            Debug.Assert(value < 10);
+            buffer[0] = (byte)('0' + value);
+        }
+
+        private void WriteValueUtf8(long value, int bytesNeeded)
+        {
```

Comment:
What can we do to Utf8Formatter to be able to use it here directly?

Reply:
We have to take a similar approach as we did for the default format, where we have a short-circuit for it: https://github.com/dotnet/corefx/pull/25424, but I am not sure how feasible it will be for a general purpose formatter.

Currently, we have the overhead of a switch statement for the 'D' Standard Format and then we have unnecessary checks based on the precision (which is default for our JsonWriter, i.e. NoPrecision).

It might be sufficient to keep the switch statement, but add an `if (precision.IsDefault)` short-circuit in all the implementations. That alone might get us most of the way there.

```C#
        TryFormat(long value, Span<byte> destination, out int bytesWritten, StandardFormat format = default)
            => TryFormatInt64(value, 0xffffffffffffffff, destination, out bytesWritten, format);


        private static bool TryFormatInt64(long value, ulong mask, Span<byte> destination, out int bytesWritten, StandardFormat format)
        {
            if (format.IsDefault)
            {
                  ...
            }
            switch (format.Symbol)
            {
                  ...
                  case 'D':
                    return TryFormatInt64D(value, format.Precision, destination, out bytesWritten);
            }
        ...

        private static bool TryFormatInt64D(long value, byte precision, Span<byte> destination, out int bytesWritten)
        {
            ...
            return TryFormatUInt64D((ulong)value, precision, destination, insertNegationSign, out bytesWritten);
        }

        private static bool TryFormatUInt64D(ulong value, byte precision, Span<byte> destination, bool insertNegationSign, out int bytesWritten)
        {
            int leadingZeroCount = ((precision == StandardFormat.NoPrecision) ? 0 : (int)precision) - digitCount;
            if (leadingZeroCount < 0)
            {
                leadingZeroCount = 0;
            }

            int requiredBufferLength = digitCount + leadingZeroCount;
```

Reply:
We should consider exposing TryFormatInt64D (and others). If we need it, our customers will too.
cc: @GrabYourPitchforks